src/apps/lwaftr/fragmentv4_test.lua

```diff
@@ -13,19 +16,32 @@ local ffi = require("ffi")
 -- Returns a new packet, which contains an Ethernet frame, with an IPv4 header,
 -- followed by a payload of "payload_size" random bytes.
 --
-local function make_ipv4_packet(payload_size)
+local function make_ipv4_packet(payload_size, vlan_tag)
+   local eth_size = eth_proto:sizeof()
+   if vlan_tag and vlan_tag > 0 then
+      eth_size = eth_size + 4  -- VLAN tag takes 4 extra bytes
+   else
+      vlan_tag = nil
+   end
    local pkt = packet.allocate()
-   pkt.length = eth_proto:sizeof() + ip4_proto:sizeof() + payload_size
+   pkt.length = eth_size + ip4_proto:sizeof() + payload_size
    local eth_header = eth_proto:new_from_mem(pkt.data, pkt.length)
-   local ip4_header = ip4_proto:new_from_mem(pkt.data + eth_header:sizeof(),
-                                             pkt.length - eth_header:sizeof())
-   assert(pkt.length == eth_header:sizeof() + ip4_header:sizeof() + payload_size)
+   local ip4_header = ip4_proto:new_from_mem(pkt.data + eth_size,
+                                             pkt.length - eth_size)
+   assert(pkt.length == eth_size + ip4_header:sizeof() + payload_size)
 
    -- Ethernet header. The leading bits of the MAC addresses are those for
    -- "Intel Corp" devices, the rest are arbitrary.
    eth_header:src(eth_proto:pton("5c:51:4f:8f:aa:ee"))
    eth_header:dst(eth_proto:pton("5c:51:4f:8f:aa:ef"))
-   eth_header:type(0x0800) -- IPv4
+
+   if vlan_tag then
+      eth_header:type(0x8100) -- VLAN TPID
+      wr16(pkt.data + eth_proto:sizeof(), vlan_tag)
```

Comment:
I think the vlan_tag variable would be better named "vlan_id", to correspond with the spec language.  Otherwise "tag" is ambiguous as to whether it includes the 0x81000 header or no.